### PR TITLE
[Scheduler Pattern] (Add) scheduler pattern #76

### PR DESCRIPTION
**Pull request by tiennm99**
I just implemented common and easy to understand disciplines like: First Come First Served, Priority scheduling, Shortest remaining time first, Round-robin scheduling.

It would be great if someone were generous enough to implement more complicated disciplines.

### CODE REVIEW

File: scheduler/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+---
+title: Scheduler Pattern
+category: Creational
+language: en
+tag:
```
Comment (iluwatar) [inline scheduler/README.md:5]: Please pick suitable tags from https://github.com/iluwatar/java-design-patterns/wiki/07.-Categories-and-Tags

File: scheduler/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+---
+
+## Name
+Scheduler Design Pattern
+
+## Intent
+The Scheduler Design Pattern is used to manage and coordinate the execution of tasks or jobs in a system. It provides a mechanism for scheduling and executing tasks at specific times, intervals, or in response to certain events. This pattern is especially useful when dealing with asynchronous operations, background processing, and resource allocation.
+
+## Explanation
+The Scheduler Design Pattern is designed to decouple task scheduling from the actual execution of tasks. It abstracts the scheduling logic, making it possible to change or extend the scheduling behavior without affecting the tasks themselves. This pattern allows for efficient resource utilization, load balancing, and prioritization of tasks.
```
Comment (iluwatar) [inline scheduler/README.md:15]: The explanation section consists of multiple subsections. See https://github.com/iluwatar/java-design-patterns/wiki/02.-Pattern-template

File: scheduler/pom.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        <version>1.26.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>scheduler</artifactId>
+
+    <properties>
+        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.target>17</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
```
Comment (iluwatar) [inline scheduler/pom.xml:18]: No need to set these here, they come from the parent pom.xml

File: scheduler/pom.xml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>5.9.2</version>
```
Comment (iluwatar) [inline scheduler/pom.xml:23]: The version property can be left out and use the latest

File: scheduler/src/test/java/com/iluwatar/scheduler/FirstComeFirstServedSchedulerTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    assertEquals(5, taskCompletedOrder.size());
+    assertIterableEquals(List.of(1, 2, 3, 4, 5), taskCompletedOrder.keySet());
+    assertEquals(3, taskCompletedOrder.get(1));
+    assertEquals(5, taskCompletedOrder.get(2));
+    assertEquals(10, taskCompletedOrder.get(3));
+    assertEquals(12, taskCompletedOrder.get(4));
+    assertEquals(13, taskCompletedOrder.get(5));
+  }
+}
```
Comment (iluwatar) [inline scheduler/src/test/java/com/iluwatar/scheduler/FirstComeFirstServedSchedulerTest.java:49]: Best practise is to create App.java with main function to run the example (see the other patterns). I think we can move this class under tests then.

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by tiennm99** [comment]
Gotcha, I'll fix them later. Maybe on next weekend.

**Comment by tiennm99** [comment]
I fixed problems in your comments. Please check them again @iluwatar

**Comment by iluwatar** [comment]
Please check the failing build and conflicts

**Comment by tiennm99** [comment]
Sorry I'm too busy these days. I will fix it by the end of Jan 2024.

**Comment by tiennm99** [comment]
I have created new PR #2794

**Pull request closed without merging**